crates/walrus-service/src/client/cli/backfill.rs

```diff
@@ -213,49 +283,90 @@ pub(crate) async fn run_blob_backfill(
         .open(&pushed_state)
         .context("opening pushed state file")?;
 
-    // Ingest via stdin, then process the blob_ids that have been stored in `backfill_dir`.
-    for line in std::io::stdin().lines() {
-        let line = line?;
-        let Ok(blob_id): Result<BlobId, _> = line.trim().parse() else {
-            tracing::error!(?line, "Failed to parse blob ID from stdin");
-            continue;
-        };
-        let blob_filename = backfill_dir.join(blob_id.to_string());
+    // Read all the files in the backfill dir, and try to backfill all of them before reading the
+    // directory again.
+    loop {
+        let entries = std::fs::read_dir(&backfill_dir)?
+            .filter_map(|entry| {
+                entry
+                    .inspect_err(|error| {
+                        tracing::error!(?error, "error reading directory entry; skipping...")
+                    })
+                    .ok()
+                    .map(|entry| entry.path())
+            })
+            .collect::<Vec<_>>();
 
-        match std::fs::read(&blob_filename) {
-            Ok(blob) => {
+        for blob_filename in entries.iter() {
+            process_file_and_backfill(
+                &client,
+                &node_ids,
+                blob_filename,
+                &mut pushed_blobs,
+                &mut pushed_state,
+            )
+            .await;
+        }
+
+        // Small delay to avoid constant reads if the directory is empty.
+        if entries.is_empty() {
+            tracing::info!("no more blobs to backfill; sleeping for 500ms");
```

Comment:
What would be the stop signal for the write job? Assuming the read job is much faster, we could use an empty directory as the signal. Or have a timeout where it stops after the directory is empty for some time.